include/ethernet.h

```diff
@@ -0,0 +1,91 @@
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#ifndef DOSBOX_ETHERNET_H
+#define DOSBOX_ETHERNET_H
+
+#include "config.h"
+#include <functional>
+
+/** A virtual Ethernet connection
+ * While emulated Ethernet adapters provide the ability for the guest OS to
+ * send and receive Ethernet packets, the emulator itself needs to pass these
+ * packets to device on the host system.
+ * The EthernetConnection class provides a virtual connection for passing
+ * packets to and from a backend, intended for use by emulated adapter code.
+ */
+class EthernetConnection
+{
+    public:
+        /** Initializes the connection.
+         * After creating an EthernetConnection, it must have its connection
+         * opened and initialized before further use.
+         * This is done in a separate function instead of a constructor so any
+         * error code may be propagated.
+         * Do not call this function twice.
+         * @return True if the connection is initialized, false otherwise
+         */
+        virtual bool Initialize(void) = 0;
+
+        /** Closes the connection.
+         * Owners of an EthernetConnection are expected to delete it when
+         * finished with the connection.
+         * The destructor will close the connection and free any
+         * per-connection resources.
+         */
+        virtual ~EthernetConnection(void)
+        {
+        }
+
+        /** Sends a packet through the connection.
+         * This function makes no guarantees as to whether the packet actually
+         * gets stored or sent anywhere. Errors are not propagated from
+         * the concrete implementation of the connection.
+         * This matches the lossy nature of sending packets through hardware.
+         * @param packet A pointer to bytes of an Ethernet frame
+         * @param size The size (in bytes) of the Ethernet frame
+         */
+        virtual void SendPacket(uint8_t* packet, int size) = 0;
+
+        /** Gets all pending packets from the connection.
+         * This function passes each pending packet to the callback function.
+         * The callback provides a pointer to the bytes of an Ethernet frame,
+         * and the size (in bytes) of the Ethernet frame.
+         * Using a lambda for std::function allows easy looping over GetPackets
+         * instead of jumping around between functions and sharing global state.
+         * This function does not define the state of the packet data outside
+         * the callback. Copy the packet data if you need to use it later.
+         * @param callback The function called for each pending packet
+         */
+        virtual void GetPackets(std::function<void(uint8_t*, int)> callback) = 0;
```

Comment:
Most network calls indicate success or failure or quantity.
 - `socket()` creation returns an id or `-1` on failure
 - `bind()` returns -1 on failure
 - `connect()` returns -1 on failure
 - `send()` returns number of bytes transmitted or -1 on failure
 - `recv()` returns number of bytes transmitted or -1 on failure
 - `close()` returns -1 on failure

But I realize we're not directly interfacing with TCP/IP, and uses pcap and slirp calls.
Do their `send()` and `receive()` equivalent calls not return anything?